### DISCUSSION TITLE
How to get the value out of Hy function?

### DISCUSSION THREAD

**Discussion by pauljurczak** [hylang/hy#2458] [Q&A] [answered]
I would like to evaluate functions defined in Hy inside Python code and get the return values given the argument values. For example:
```
import hy

s = '(setv x 9)(defn incr [x] (+ x 1))(incr x)'
v = hy.eval(hy.read(s))
print(v)  
```
`(setv x 9)(defn incr [x] (+ x 1))(incr x)` produces `10` when fed to Hy REPL, but produces `None` in above script. What do I have to change?

**Comment by Kodiologist** [comment, marked as answer]
Edit for future readers: you need `hy.read_many` to read multiple forms in a single string at one go. `hy.read` reads only the first form.

Did you look at the docstring of `hy.read`?

**Reply by pauljurczak** [reply]
I looked at https://docs.hylang.org/en/stable/interop.html#using-hy-from-python: *To execute Hy code from a string, use hy.read() to convert it to [models](https://docs.hylang.org/en/stable/syntax.html#models) and then hy.eval() to evaluate it.*

**Reply by Kodiologist** [reply]
That's not the docstring of `hy.read`.

**Reply by pauljurczak** [reply]
I'm looking at https://docs.hylang.org/en/stable/api.html?highlight=read#hy.read and I don't understand the description. It seems to refer to Hy code not the Python API.

**Reply by Kodiologist** [reply]
That text looks pretty clear to me. The fact that it refers to Hy code makes sense because Hy code is what `hy.read` reads. Hy's "Python API" is just a set of Python functions and classes, the same ones you can call from Hy. So there isn't a Hy version of `hy.read` separate from the Python version; it's the same function.

If you're still having trouble understanding the text, try explaining what confuses you about it and I'll try to clarify. I know that the documentation still needs a lot of work.

**Reply by pauljurczak** [reply]
I just found about Hy, so I'm a completely new user. I would appreciate an edit of my example to demonstrate how it's done. I've read https://docs.hylang.org/en/stable/api.html?highlight=read#hy.read several times now and I don't see why my script is not working. I also tried:
```
s = '`(setv x 9)(defn incr [x] (+ x 1))(incr x)'
```
unsuccessfully.

**Reply by Kodiologist** [reply]
Spoon-feeding you answers doesn't help you learn the language and it doesn't help improve the manual. Help me help you understand the documentation, so then I can make any needed improvements. If you still can't figure it out, then I'll be happy to tell you the answer.

**Reply by pauljurczak** [reply]
I'm assuming that you are making this conversation difficult for me, in order to improve the documentation. Given that, there is no confusion reading https://docs.hylang.org/en/stable/api.html?highlight=read#hy.read, but it introduces number of terms, which are obscure for first time user, who doesn't know Lisp, e.g. *only one form is read, and shebangs are forbidden*. I have no idea what *form* and *shebang* is.

If you want to improve the documentation for newbies, please add an example to https://docs.hylang.org/en/stable/interop.html#using-hy-from-python. I think most people would read it the same way I did and write a similar non-working script.

**Reply by pauljurczak** [reply]
Also, my use case is to evaluate a simple Lisp-like function produced by solvers like cvc5. The example I gave is representative. I don't care about learning more Lisp than that. All of my Hy coding will be this one-liner with different functions and arguments in place of `f` and `x`. I would rather not climb the learning curve for this.

**Reply by Kodiologist** [reply]
> I'm assuming that you are making this conversation difficult for me, in order to improve the documentation.

That and to help you learn the language. For example, now that I see you don't know what a form is, then that explains a lot of the difficulty. I'd recommend beginning with the tutorial, which explains what forms are, as well as many other basic concepts. The chapter on syntax is also helpful for explaining some of the details, such as the difference between forms and models.

> …please add an example to https://docs.hylang.org/en/stable/interop.html#using-hy-from-python

I agree. I think I should do that.

> Also, my use case is to evaluate a simple Lisp-like function produced by solvers like cvc5. … I would rather not climb the learning curve for this.

I'm not familiar with solvers of this kind. If I understand correctly, Hy is likely to be vast overkill for simple arithmetic, and you'll be happier using your favorite Scheme implementation, or writing a minimal Lispy expression parser for the purpose. But if Hy works for you, more power to you.

Since you've been a good sport, here's your answer: use `hy.read_many` instead of `hy.read`. Before asking any further questions, though, you should make a good-faith effort to climb that learning curve.

**Comment by pauljurczak** [comment]
> favorite Scheme implementation

I know none. Any recommendations? I've spent at least an hour searching for Lisp S-expressions evaluation in Python with very muddy results.

**Reply by Kodiologist** [reply]
Not really; I mostly use Hy and a little Emacs Lisp, and I only know Scheme by reputation. I'm not sure if there are any Schemes in Python, but I know there are a lot of mature Scheme implementations, and you could farm out evaluation to them. The inter-process communication could be slow, but Python isn't exactly a speed demon, either (I doubt that numerical solvers with Python interfaces are themselves written in pure Python), so the tradeoffs are unclear.

**Reply by pauljurczak** [reply]
Thanks. I will stay with Hy, since it works for my use case.
